### PR TITLE
replace subroutineCall and kernelCall with functionCall in ANTLR

### PR DESCRIPTION
**Pull request by hhorii**
Currently the parser distinguishes calls of subroutine and kernel.

```
kernelCall
    : Identifier LPAREN expressionList? RPAREN
    ;

subroutineCall
    : Identifier ( LPAREN expressionList? RPAREN )? indexIdentifierList
    ;
```

First, this `subroutineCall` cannot parse a subroutine call that does not take any qubit arguments (because `indexIdentifierList` has at least one element).

Second, if `indexIdentifierList` of `subroutineCall` is optional, no difference between `kernelCall` and `subroutineCall`.

This PR replaces above two definitions with a new definition `functionCall`:
```
functionCall
    : Identifier ( LPAREN expressionList? RPAREN )? indexIdentifierList?
    ;
```

### CODE REVIEW

**Comment by zachschoenfeld33** [comment]
Looks good. Maybe you could call `functionCall` something like `kernSubCall`? So this way it relates to the vocabulary of OpenQASM3 (since we don't call them "functions").

It's also true that this could get mismatched w/ gate, but I think that's just an ambiguity of the language (which already existed prior to this PR).

**Comment by stevenheidel** [comment]
We resolved this ambiguity in some last-minute discussions on the paper.

Subroutine calls take qubit arguments within the parenthesis. This is because the subroutine will not be broadcast across an array of qubits like is the case with gates.

See the example in section 3.3 of the paper:
` def segment(qubit[2] anc, qubit psi) -> bit[2] { `

**Comment by stevenheidel** [comment]
Somewhat related: https://github.com/Qiskit/openqasm/issues/185

**Comment by hhorii** [comment]
@stevenheidel thanks. I will close this PR and send another PR to support the paper's syntax.

File: source/grammar/qasm3.g4 [hunk 1]
```diff
@@ -447,27 +455,22 @@ kernelDeclaration
     : 'kernel' Identifier ( LPAREN classicalTypeList? RPAREN )? returnSignature? SEMICOLON
     ;
 
-// if have kernel w/ out args, is ambiguous; may get matched as identifier
-kernelCall
-    : Identifier LPAREN expressionList? RPAREN
+// if have function call w/ out args, is ambiguous; may get matched as identifier
+kernelOrSubroutineCall
+    : Identifier ( LPAREN expressionList? RPAREN )? indexIdentifierList?
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:460]: I'm not sure this is correct. I think we need a list which allows either expression or index identifier and put that list into the parens. Something like `exprAndIndexIdList`

Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4:460]: The test to add would be something like `sub_call(a, b, q[0] || q[1])` and see if that works...

File: source/grammar/tests/outputs/sub_and_kern_call.yaml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # indent w/ 2 spaces
 source: |
-  bit x = sub_call(10, "01") q1[0], q2;
+  bit x = sub_call(10, "01", q1[0], q2);
```
Comment (zachschoenfeld33) [inline source/grammar/tests/outputs/sub_and_kern_call.yaml:3]: Can you add index identifier specifics such as either `||` or `[:]` syntax to the test?

**Review by zachschoenfeld33 — COMMENTED** [review]
Had one comment to fix...

File: source/grammar/qasm3.g4 (hunk 1 above)
Comment (hhorii) [reply source/grammar/qasm3.g4:460]: This was wrong... As @stevenheidel mentioned, `qubit` arguments must be in parenthesis in any calls. 

```
kernelOrSubroutineCall
    : Identifier ( LPAREN expressionList? RPAREN )?
    ;
```

Then, `sub_call(a, b, q[0] || q[1])` is parsed as

```
kernelOrSubroutineCall
  sub_call
  (
  expressionList
    expression
      expressionTerminator
        a
    ,
    expression
      expressionTerminator
        b
    ,
    expression
      expression
        expressionTerminator
          expressionTerminator
            q
          [
          expression
            expressionTerminator
              0
          ]
      ||
      logicalAndExpression
        bitOrExpression
          xOrExpression
            bitAndExpression
              equalityExpression
                comparisonExpression
                  bitShiftExpression
                    additiveExpression
                      multiplicativeExpression
                        expressionTerminator
                          expressionTerminator
                            q
                          [
                          expression
                            expressionTerminator
                              1
                          ]
```

Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4]: Right--that's why I am suggesting a list that can either include expressions or index identifiers, which would be in the parenthesis. Something like:
```
exprOrIndexId : expression | indexIdentifier;
exprOrIndexIdList : (exprOrIndexId COMMA)* exprOrIndexId
```

Comment (zachschoenfeld33) [reply source/grammar/qasm3.g4]: I'm not sure we should add `||`, `:` to expression as they are specific expressions only allowed for registers. If we added to expression you could do `int x = [2:3]` which clearly doesn't make sense.

**Comment by stevenheidel** [comment]
@hhorii @zachschoenfeld33 - are we okay with pulling in @aspcompiler 's fix?

**Comment by zachschoenfeld33** [comment]
Yes works for me...

**Comment by hhorii** [comment]
@aspcompiler Thank you for your help and I'm sorry that my response is very late. I cherry-picked your commit.

File: source/grammar/tests/outputs/sub_and_kern_call.yaml (hunk 2 above)
Comment (hhorii) [reply source/grammar/tests/outputs/sub_and_kern_call.yaml]: I'm not sure that such operators are allowed excepting alias statements.

**Comment by blakejohnson** [comment]
@stevenheidel is this good to go, now?

File: source/grammar/qasm3.g4 [hunk 3]
```diff
@@ -464,7 +464,7 @@ externDeclaration
     ;
 
 // if have function call w/ out args, is ambiguous; may get matched as identifier
-kernelOrSubroutineCall
+externOrSubroutineCall
     : Identifier ( LPAREN expressionList? RPAREN )?
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:468]: Are parentheses required in the case of no arguments? I would argue that they should be...

File: source/grammar/qasm3.g4 [hunk 4]
```diff
@@ -455,27 +463,22 @@ externDeclaration
...
+// if have function call w/ out args, is ambiguous; may get matched as identifier
+externOrSubroutineCall
+    : Identifier ( LPAREN expressionList? RPAREN )?
     ;
 
 /*** Subroutines ***/
 
 subroutineDefinition
-    : 'def' Identifier ( LPAREN classicalArgumentList? RPAREN )? quantumArgumentList?
+    : 'def' Identifier ( LPAREN anyTypeArgumentList? RPAREN )?
```
Comment (zachschoenfeld33) [inline source/grammar/qasm3.g4:474]: Same comment about whether parentheses should be required in the case of no arguments...

**Review by zachschoenfeld33 — COMMENTED** [review]
One minor question: Should we require parentheses on `extern`'s and `subroutine`'s in the case of no arguments? I suggest that we do, as this will match classical function syntax and avoid any potential ambiguities...

File: source/grammar/qasm3.g4 (hunk 3 above)
Comment (aspcompiler) [reply source/grammar/qasm3.g4:468]: +1 for parentheses to remove the ambiguity with indexIdentifier: Identifier ( LBRACKET expressionList RBRACKET )?

Comment (blakejohnson) [reply source/grammar/qasm3.g4:468]: I would be okay making parentheses mandatory.

**Review by zachschoenfeld33 — APPROVED** [review]
LGTM

**Review by taalexander — APPROVED** [review]

**Pull request merged**